pkg/build/controller/controller.go

```diff
@@ -299,65 +299,66 @@ func (bc *BuildPodController) HandlePod(pod *kapi.Pod) error {
-		if secret := build.Spec.Output.PushSecret; secret != nil && currentReason != buildapi.StatusReasonMissingPushSecret {
+		currentPhase = buildapi.BuildPhasePending
+		currentReason, currentMessage = resetReasonAndMessage(previousPhase, currentPhase, previousReason, previousMessage)
+		if secret := build.Spec.Output.PushSecret; secret != nil && previousReason != buildapi.StatusReasonMissingPushSecret {
 			if _, err := bc.SecretClient.Secrets(build.Namespace).Get(secret.Name); err != nil && errors.IsNotFound(err) {
-				build.Status.Reason = buildapi.StatusReasonMissingPushSecret
-				build.Status.Message = buildapi.StatusMessageMissingPushSecret
-				glog.V(4).Infof("Setting reason for pending build to %q due to missing secret %s/%s", build.Status.Reason, build.Namespace, secret.Name)
+				currentReason = buildapi.StatusReasonMissingPushSecret
+				currentMessage = buildapi.StatusMessageMissingPushSecret
+				glog.V(4).Infof("Setting reason for pending build to %q due to missing secret %s/%s", currentReason, build.Namespace, secret.Name)
 			}
 		}
 
 	case kapi.PodSucceeded:
-		build.Status.Reason = ""
-		build.Status.Message = ""
-		// Check the exit codes of all the containers in the pod
-		nextStatus = buildapi.BuildPhaseComplete
+		currentPhase = buildapi.BuildPhaseComplete
+		currentReason, currentMessage = resetReasonAndMessage(previousPhase, currentPhase, previousReason, previousMessage)
+		//Check the exit code for all containers in the pod
 		if len(pod.Status.ContainerStatuses) == 0 {
 			// no containers in the pod means something went badly wrong, so the build
 			// should be failed.
 			glog.V(2).Infof("Failing build %s/%s because the pod has no containers", build.Namespace, build.Name)
-			nextStatus = buildapi.BuildPhaseFailed
+			currentPhase = buildapi.BuildPhaseFailed
 		} else {
 			for _, info := range pod.Status.ContainerStatuses {
 				if info.State.Terminated != nil && info.State.Terminated.ExitCode != 0 {
-					nextStatus = buildapi.BuildPhaseFailed
+					currentPhase = buildapi.BuildPhaseFailed
 					break
 				}
 			}
 		}
 
 	case kapi.PodFailed:
-		nextStatus = buildapi.BuildPhaseFailed
+		currentPhase = buildapi.BuildPhaseFailed
+		currentReason, currentMessage = resetReasonAndMessage(previousPhase, currentPhase, previousReason, previousMessage)
 
 	default:
-		build.Status.Reason = ""
-		build.Status.Message = ""
+		currentReason, currentMessage = resetReasonAndMessage(previousPhase, currentPhase, previousReason, previousMessage)
 	}
 
-	// Update the build object when it progress to a next state or the reason for
-	// the current state changed.
-	if (!hasBuildPodNameAnnotation(build) || build.Status.Phase != nextStatus) && !buildutil.IsBuildComplete(build) {
+	// Update the build object when it:
+	// - Progresses to the next phase
+	// - The reason has changed
+	if (!hasBuildPodNameAnnotation(build) || previousPhase != currentPhase || previousReason != currentReason) && !buildutil.IsBuildComplete(build) {
```

Comment:
|| previousMessage != currentMessage  (yes i know, the Reason check probably covers this, but one day it might not)